### PR TITLE
Finalize support for Login and Credentials

### PR DESCRIPTION
**Pull request by cmelchior**
This PR finalizes login functionality and RealmCredentials + JavaDoc
Deleted old CredentialsTest and created RealmCredentialsTests instead.

TODO
- [ ] Implemenet missing support for Custom Functions login
- [ ] Implement missing support for Api Keys.

### CODE REVIEW

File: realm/realm-library/src/objectServer/java/io/realm/RealmCredentials.java [hunk 1]
```diff
@@ -19,139 +19,148 @@
...
      * FIXME
+     * <p>
+     * This provider must be enabled on MongoDB Realm to work.
+     *
+     * @return a set of credentials that can be used to log into MongoDB Realm using
+     * {@link RealmApp#loginAsync(RealmCredentials, RealmApp.Callback)}.
      */
     public static RealmCredentials customFunction(String functionName, Object... arguments) {
-//        assertStringNotEmpty(idToken, "idToken");
+        // Util.checkEmpty(idToken, "idToken");
```
Comment (nhachicha) [inline realm/realm-library/src/objectServer/java/io/realm/RealmCredentials.java:106]: FIXME for any upcoming params checks

File: realm/realm-library/src/objectServer/java/io/realm/RealmCredentials.java [hunk 2]
```diff
@@ -19,139 +19,148 @@
...
     /**
-     * FIXME
-     * Creates credentials based on a Google login.
+     * Creates credentials representing a login using an Google access token.
+     * <p>
+     * This provider must be enabled on MongoDB Realm to work.
      *
-     * @param googleToken a google userIdentifier acquired by logging into Google.
-     * @return a set of credentials that can be used to log into the Object Server using
+     * @param googleToken the access token returned when logging in to Facebook.
```
Comment (nhachicha) [inline realm/realm-library/src/objectServer/java/io/realm/RealmCredentials.java:143]:
```suggestion
     * @param googleToken the access token returned when logging in to Google.
```

File: realm/realm-library/src/objectServer/java/io/realm/RealmCredentials.java [hunk 3]
```diff
@@ -173,64 +182,34 @@ public String asJson() {
...
-    }
-
     private RealmCredentials(OsAppCredentials credentials) {
         this.osCredentials = credentials;
     }
 
     /**
-     * FIXME
+     * This enum contains the list of identity providers supported by MongoDB Realm.
+     * All of these expect {@link #EMAIL_PASSWORD} must be enabled manually on MongoDB Realm to
```
Comment (nhachicha) [inline realm/realm-library/src/objectServer/java/io/realm/RealmCredentials.java:191]:
```suggestion
     * All of these except {@link #EMAIL_PASSWORD} must be enabled manually on MongoDB Realm to
```

File: realm/realm-library/src/objectServer/java/io/realm/RealmApp.java [hunk 4]
```diff
@@ -153,6 +154,80 @@ public RealmUser switchUser(RealmUser user) {
...
+     * <pre>
+     * {@code
+     * // Example
+     * RealmApp app = new RealmApp("app-id")
+     * RealmUser user = app.login(RealmCredentials.anonymous());
+     * app.linkUser(RealmCredentials.emailPassword("email", "password"));
+     * }
+     * </pre>
+     * <p>
+     * Note: It is not possible to link two existing users of MongoDB Realm. The provided credentials
```
[the remark this answers is not in the document]
Comment (cmelchior) [reply realm/realm-library/src/objectServer/java/io/realm/RealmApp.java:214]: No, you can also link e.g. facebook credentials to a user created with email/password. But I think the primary use case is upgrading "demo"-users.

**Pull request merged**